pkg/cluster/streams.go

```diff
@@ -0,0 +1,261 @@
+	"reflect"
+	"strings"
+
+	acidv1 "github.com/zalando/postgres-operator/pkg/apis/acid.zalan.do/v1"
+	zalandov1alpha1 "github.com/zalando/postgres-operator/pkg/apis/zalando.org/v1alpha1"
+	"github.com/zalando/postgres-operator/pkg/util"
+	"github.com/zalando/postgres-operator/pkg/util/config"
+	"github.com/zalando/postgres-operator/pkg/util/constants"
+	"github.com/zalando/postgres-operator/pkg/util/k8sutil"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var outboxTableNameTemplate config.StringTemplate = "{table}_{eventtype}_outbox"
+
+func (c *Cluster) createStreams() error {
+	c.setProcessName("creating streams")
+
+	fes := c.generateFabricEventStream()
+	_, err := c.KubeClient.FabricEventStreams(c.Namespace).Create(context.TODO(), fes, metav1.CreateOptions{})
+	if err != nil {
+		return fmt.Errorf("could not create event stream custom resource: %v", err)
+	}
+
+	return nil
+}
+
+func (c *Cluster) updateStreams(newEventStreams *zalandov1alpha1.FabricEventStream) error {
+	c.setProcessName("updating event streams")
+
+	_, err := c.KubeClient.FabricEventStreams(newEventStreams.Namespace).Update(context.TODO(), newEventStreams, metav1.UpdateOptions{})
+	if err != nil {
+		return fmt.Errorf("could not update event stream custom resource: %v", err)
+	}
+
+	return nil
+}
+
+func (c *Cluster) deleteStreams() error {
+	c.setProcessName("deleting event streams")
+
+	// check if stream CRD is installed before trying a delete
+	_, err := c.KubeClient.CustomResourceDefinitions().Get(context.TODO(), constants.EventStreamSourceCRDName, metav1.GetOptions{})
+	if k8sutil.ResourceNotFound(err) {
+		return nil
+	}
+
+	err = c.KubeClient.FabricEventStreams(c.Namespace).Delete(context.TODO(), c.Name, metav1.DeleteOptions{})
+	if err != nil {
+		return fmt.Errorf("could not delete event stream custom resource: %v", err)
+	}
+
+	return nil
+}
+
+func (c *Cluster) syncPostgresConfig() error {
+
+	desiredPostgresConfig := make(map[string]interface{})
+	slots := make(map[string]map[string]string)
+
+	// if streams are defined wal_level must be switched to logical and slots have to be defined
+	desiredPostgresConfig["postgresql"] = map[string]interface{}{patroniPGParametersParameterName: map[string]string{"wal_level": "logical"}}
+
+	for _, stream := range c.Spec.Streams {
+		slotName := c.getLogicalReplicationSlot(stream.Database)
+
+		if slotName == "" {
+			slot := map[string]string{
+				"database": stream.Database,
+				"plugin":   "wal2json",
+				"type":     "logical",
+			}
+			slots[constants.EventStreamSourceSlotPrefix+"_"+stream.Database] = slot
+		}
+	}
+
+	if len(slots) > 0 {
+		c.logger.Debugf("setting wal level to 'logical' in Postgres configuration")
+		for slotName, slot := range slots {
+			c.logger.Debugf("creating logical replication slot %q in database %q", slotName, slot["database"])
+		}
+		desiredPostgresConfig["slots"] = slots
+	} else {
+		return nil
+	}
+
+	pods, err := c.listPods()
+	if err != nil || len(pods) == 0 {
+		return err
```

Comment:
is it intended to return `nil` in case `len(pods) == 0` ?
that is not necessary because the combination of [line 95](https://github.com/zalando/postgres-operator/pull/1570/files#diff-6592d6884b6c11bdc6a11828daa2592c9c23b29865c49a30a55f6be5c7602ff7R95) and [line 110](https://github.com/zalando/postgres-operator/pull/1570/files#diff-6592d6884b6c11bdc6a11828daa2592c9c23b29865c49a30a55f6be5c7602ff7R110) already does it

Reply:
returning a warning now